### PR TITLE
Modularity: Hook invocation Process [Entrypoint Stage]

### PR DESCRIPTION
**Pull request by legendko**
1. Added implementation for module/hook invocation. Complete flow is ready for the `Entrypoint stage` for `/openrtb2/auction` and `/openrtb2/amp` endpoints. The execution of stages consists of multiple generic functions which will allow to reuse the invocation process for the next stages.
2. The modules may `reject` the request entirely returning an `empty bid response` with corresponding `NBR code`.
3. `StageOutcome` will be used for storing the details of the modules' execution results and in case debug is enabled they will be added to the response.

### CODE REVIEW

File: endpoints/openrtb2/amp_auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction.go:147]: I recommend adding a comment explaining there is no body for AMP requests to explain the nil argument. Or perhaps use a private package scoped variable called `nilBody` set to nil and pass that in here.

File: endpoints/openrtb2/amp_auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction.go:234]: The `FindReject` function is allocating memory just to check the type, which is not ideal for performance. I also don't believe it needs to return a pointer. Please implement something like a new `CastRejectErr` method:

```
func FindFirstRejectOrNil(errors []error) RejectError {
	for _, err := range errors {
		if rejectErr, ok := CastRejectErr(err); ok {
			return rejectErr
		}
	}
	return nil
}

func CastRejectErr(err error) (RejectError, bool) {
    return err.(RejectError)
}
```

and update the call side to:

```
rejectErr := hookexecution.CastRejectErr(err)
```

File: endpoints/openrtb2/amp_auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction.go:272]: You can simplify this with: `NBR: reject.NBR.Ptr()`

File: endpoints/openrtb2/amp_auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction.go:348]: You can remove an `if` statement by combining the top two conditions:
`if reqWrapper != nil  && reqWrapper.Test == 1 && eRErr == nil {`

File: endpoints/openrtb2/amp_auction_test.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction_test.go:1943]: Please always include labels on object initializers.

File: endpoints/openrtb2/amp_auction_test.go
Comment (SyntaxNode) [inline endpoints/openrtb2/amp_auction_test.go:1956]: There is no need to use a table driven test if there is only 1 condition. Will there be more conditions in the near future?

File: endpoints/openrtb2/auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:236]: Please use the same new reject method here as in AMP to avoid allocating an array/slice just to perform the type check.

File: endpoints/openrtb2/auction.go
Comment (SyntaxNode) [inline endpoints/openrtb2/auction.go:341]: Why are we trying to unmarshal if the module has issues a rejection?

File: endpoints/openrtb2/auction_test.go
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:4662]: Please always include labels on object initializers.

File: endpoints/openrtb2/auction_test.go
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:4673]: Please always include labels on object initializers. If there is only one test case, please remove the table driven structure.

File: endpoints/openrtb2/auction_test.go
Comment (SyntaxNode) [inline endpoints/openrtb2/auction_test.go:4678]: Good use of `t.Run`. That's something we're missing from most of our tests. Glad to see if being added in by your tests.

File: hooks/hookexecution/errors.go
Comment (SyntaxNode) [inline hooks/hookexecution/errors.go:13]: There is no need to use `Sprint` in this case.

File: hooks/hookexecution/errors.go
Comment (SyntaxNode) [inline hooks/hookexecution/errors.go:20]: There is no need to use `Sprint` in this case.

File: hooks/hookexecution/execution.go
Comment (SyntaxNode) [inline hooks/hookexecution/execution.go:75]: Why is the `resp` channel being closed here? This introduces a race condition where `resp` channel would be closed before the `executeHook` have finished sending their results. Better to remove the wait group and add a defer to close the `resp` channel.

File: hooks/hookexecution/execution.go
Comment (SyntaxNode) [inline hooks/hookexecution/execution.go:61]: Better to rename `done` to its true purpose of `rejected`. It is not a way for `executeHook` to signal work has completed, but rather for `collectHookResponses` to exit pending `executeHook` calls early when a rejection is first encountered.

File: hooks/hookexecution/execution.go
Comment (SyntaxNode) [inline hooks/hookexecution/execution.go:95]: This case can be removed. There is an incredibly low change a rejection would've occurred at this point where the code for this optimization isn't worth it IMHO.

File: hooks/hookexecution/execution.go
Comment (SyntaxNode) [inline hooks/hookexecution/execution.go:180]: The code flow is hard to follow with all of the continue statements. What does it mean to continue within this loop? Is there a way to rewrite this loop and/or use helper methods to make the code flow more obvious?

File: hooks/hookexecution/executor.go
Comment (SyntaxNode) [inline hooks/hookexecution/executor.go:20]: What is an entity?

File: hooks/hookexecution/executor.go
Comment (SyntaxNode) [inline hooks/hookexecution/executor.go:152]: Please move the defer up one line to be directly below the `Lock` call.

File: hooks/hookstage/mutation.go
Comment (SyntaxNode) [inline hooks/hookstage/mutation.go:7]: Do we also need "MutationAdd"?

File: endpoints/openrtb2/amp_auction_test.go
Comment (4lexvav) [reply endpoints/openrtb2/amp_auction_test.go:1956]: We are going to extend this with new test cases in the following PRs, to check rejections made at other stages.

File: endpoints/openrtb2/auction.go
Comment (4lexvav) [reply endpoints/openrtb2/auction.go:341]: You are right, this makes no sense within this PR, so I've removed it.
In one of the next PRs, we are adding debug info to `BidResponse`, but only if the debug mode enabled, which is determined from `BidRequest`.

File: endpoints/openrtb2/amp_auction.go
Comment (4lexvav) [reply endpoints/openrtb2/amp_auction.go:234]: Sure, we will add `CastRejectErr` func to avoid extra memory allocation. However, since we work with the concrete type `RejectError`, we are forced to return a pointer if we want to return `nil` when `RejectError` is not found.

File: endpoints/openrtb2/auction_test.go
Comment (4lexvav) [reply endpoints/openrtb2/auction_test.go:4673]: We will add new test cases in the next PRs here too.

File: hooks/hookexecution/execution.go
Comment (4lexvav) [reply hooks/hookexecution/execution.go:75]: Perhaps I’m missing something, but in my understanding `wg.Wait()` is a blocking operation, so when `close(resp)` is called, all goroutines have already finished their execution and it’s safe to close the channel.

File: hooks/hookexecution/executor.go
Comment (4lexvav) [reply hooks/hookexecution/executor.go:20]: An entity specifies the type of object that was processed during the execution of the stage.
For most stages, the entity will be equal to one of the constant values. But for the `bidder-request` and `raw-bidder-response` stages, the entity is set to the bidder name. This is similar to what PBS-Java does.

File: endpoints/openrtb2/amp_auction.go
Comment (SyntaxNode) [reply endpoints/openrtb2/amp_auction.go]: Ah, ok. Got it. Makes sense.

File: endpoints/openrtb2/amp_auction_test.go
Comment (SyntaxNode) [reply endpoints/openrtb2/amp_auction_test.go:1954]: Gotcha. Sounds good.

File: hooks/hookexecution/execution.go
Comment (SyntaxNode) [reply hooks/hookexecution/execution.go:75]: Yes, `wg.Wait()` is blocking and will not unblock until every `executeHook` completes. Let me take another look.

File: hooks/hookexecution/executor.go
Comment (SyntaxNode) [reply hooks/hookexecution/executor.go:21]: Thank you. I think a comment will make it clear.

File: endpoints/openrtb2/amp_auction.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/amp_auction.go:265]: Nitpick: Could you re-format this function so the parameters are all on one line to follow the patterns of other functions in this file?

File: endpoints/openrtb2/amp_auction.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/amp_auction.go:280]: Nitpick: Could you re-format this function so the parameters are all on one line to follow the patterns of other functions in this file?

File: endpoints/openrtb2/auction.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction.go:252]: Can you update tests to cover these lines if possible?

File: endpoints/openrtb2/auction.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction.go:257]: Same parameter one line re-formatting comment

File: endpoints/openrtb2/auction.go
Comment (AlexBVolcy) [inline endpoints/openrtb2/auction.go:275]: Same parameter one line re-formatting comment

File: hooks/hookexecution/context.go
Comment (AlexBVolcy) [inline hooks/hookexecution/context.go:24]: Is it possible to update tests to cover this line?

File: hooks/hookexecution/context.go
Comment (AlexBVolcy) [inline hooks/hookexecution/context.go:54]: Is it possible to update tests to cover these lines?

File: hooks/hookexecution/execution.go
Comment (AlexBVolcy) [reply hooks/hookexecution/execution.go:174]: Yeah I would agree with this comment

File: hooks/hookexecution/execution.go
Comment (AlexBVolcy) [inline hooks/hookexecution/execution.go:187]: Is it possible to update tests to cover these lines?

File: hooks/hookexecution/executor.go
Comment (AlexBVolcy) [inline hooks/hookexecution/executor.go:140]: Is it necessary to have these functions defined if there is no code actually present in them yet?

File: hooks/hookexecution/execution.go
Comment (4lexvav) [reply hooks/hookexecution/execution.go:75]: Sure, we will add more tests to ensure 100% coverage.

File: hooks/hookexecution/execution.go
Comment (4lexvav) [reply hooks/hookexecution/execution.go]: We've done some refactoring of this function, please check if it's clearer now.

File: endpoints/openrtb2/amp_auction.go
Comment (4lexvav) [reply endpoints/openrtb2/amp_auction.go:265]: Please clarify, do you wish us to re-format functions to one line only in files we edited or is it a whole project convention and we should use the same format across all functions we created? Because we tried to keep the line length under 120 characters long in all functions we created.

File: endpoints/openrtb2/auction.go
Comment (4lexvav) [reply endpoints/openrtb2/auction.go:252]: Actually, in this PR we do not invoke any hooks within the `exchange.HoldAuction` method, so there will be no rejection error and this code is not unreachable right now. But it will be added in next PR with corresponding tests, so if you think it is better to get out of this code from current PR - we can remove it for now.

File: hooks/hookexecution/executor.go
Comment (4lexvav) [reply hooks/hookexecution/executor.go:140]: Yes, we can remove it from this PR and add it in subsequent PRs with implementation of these methods.

File: hooks/hookexecution/context.go
Comment (4lexvav) [reply hooks/hookexecution/context.go:24]: The purpose of this logic is for communication between different stages, so that module can pass some data to itself from earlier stages to later ones. Right now we have implementation for `entrypoint` stage only, but we will definitely cover this logic with implementations of new stages.

File: hooks/hookexecution/execution.go
Comment (4lexvav) [reply hooks/hookexecution/execution.go]: Unfortunately, right now we have implementation only for the `entrypoint` stage which is rejectable. But we are going to cover this logic with tests within the PRs for `all-processed-bid-responses` and `auction-response` stages, for which rejection is not allowed.

File: endpoints/openrtb2/amp_auction.go
Comment (bsardo) [reply endpoints/openrtb2/amp_auction.go]: Hi @4lexvav, the team discussed this offline and decided that we want to favor readability over consistency in this case so we prefer you stick with your approach of breaking long one-line function signatures up so that they span multiple lines. 120 characters seems like a good rule of thumb.
@AlexBVolcy, there are some cases where we already do this and we will continue to make this change throughout our code base over time.

File: hooks/hookexecution/execution.go
Comment (4lexvav) [reply hooks/hookexecution/execution.go:75]: We added tests to cover more cases, but right now we cannot achieve 100% code coverage because some of the logic is reachable at later stages only. But we will add more tests in PRs with new stages.

File: endpoints/openrtb2/amp_auction.go
Comment (4lexvav) [reply endpoints/openrtb2/amp_auction.go:265]: Sure, I restored the multiline format of function signatures.

File: endpoints/openrtb2/auction_test.go
Comment (bsardo) [inline endpoints/openrtb2/auction_test.go:4707]: According to the spec, a rejection in the entrypoint stage should result in a 204 or 4XX:

> For /openrtb2/auction and /openrtb2/video modules must be able to modify the original POST request that comes into the server or reject the request entirely. Rejection at this stage may result in HTTP 204 or a 4xx error and a metric associating the rejection with the module.

In that case, the body should be empty if using 204 or contain an error message if using 4XX.

What is PBS-Java doing?

File: endpoints/openrtb2/amp_auction_test.go
Comment (bsardo) [inline endpoints/openrtb2/amp_auction_test.go:1989]: According to the spec, a rejection at the entrypoint stage should result in a 204:

> For /openrtb2/amp and /setuid, modules must be able to modify the GET query string or reject the request entirely. Rejection results in HTTP 204 and a metric associating the rejection with the module.

In that case, the body should be empty if using 204.

File: hooks/hookstage/mutation.go
Comment (4lexvav) [reply hooks/hookstage/mutation.go:6]: Right now the hooks framework does not provide implementations for the `add`, `update` or `delete` operations. Instead, it all falls on the shoulders of the module that will create and return mutations from its hooks. Therefore, the module itself is responsible for choosing the right type for its mutations and defining the cases in which it should return an error. But we do not force modules to follow some algorithm for mutation types. The errors returned from mutations are logged as the warning messages and such mutations are skipped: https://github.com/prebid/prebid-server/blob/cfd8e8a8ba865def1fbabea2ceb5074a48701656/hooks/hookexecution/execution.go#L247-L254

But with introduction of module we will see more use cases and will be able to develop the `ChangeSet` concept to make it possible to provide implementations for different operations, define their behaviour and edge cases (in which case should the mutation fail or proceed?).

File: hooks/plan.go
Comment (legendko) [reply hooks/plan.go]: updated

File: hooks/hookexecution/executor.go
Comment (legendko) [reply hooks/hookexecution/executor.go]: done

File: hooks/hookexecution/outcome.go
Comment (legendko) [reply hooks/hookexecution/outcome.go]: Yes, this map will be used in the outcomes later for enriching the response. Removed from this PR.

File: hooks/hookexecution/outcome.go
Comment (legendko) [reply hooks/hookexecution/outcome.go]: Sure, we updated them with underscores. There was not any particular reason to use the flat case.

File: hooks/hookexecution/outcome_test.go
Comment (legendko) [reply hooks/hookexecution/outcome_test.go]: done

File: hooks/hookexecution/execution.go
Comment (legendko) [reply hooks/hookexecution/execution.go:120]: Correct, we use this behavior for stopping the processing ASAP according to the PRD requirements. Other hooks from this group will finish their work in their own goroutines if they already started working, but their results wont have any effect. The next groups of hooks won't even start processing and we exit immediately.

File: endpoints/openrtb2/auction_test.go
Comment (legendko) [reply endpoints/openrtb2/auction_test.go:4707]: Yes, we've read this but PBS-Java is always sending a 200 response with an empty bid response, so we decided to do the same for PBS-Go

File: endpoints/openrtb2/amp_auction_test.go
Comment (legendko) [reply endpoints/openrtb2/amp_auction_test.go:1989]: Same here

**Review by bsardo — APPROVED** [review]

**Review by AlexBVolcy — APPROVED** [review]

**Pull request merged**
